### PR TITLE
Update appearance of keyboard shortcut tooltips

### PR DESCRIPTION
**Pull request by janfaracik**
Small one in preparation for #7569. This PR adds a new component for showing keyboard shortcuts, it's currently used in the tooltip for search bars.

https://github.com/jenkinsci/jenkins/assets/43062514/24515d14-b996-4dda-8b9a-4cd27b798b55

This will be used for the Command Palette button too, where hovering over the Search icon will list what shortcuts activate it.

### Testing done

* Shortcut and animation display as expected

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@jenkinsci/sig-ux 

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8435"></a>

### CODE REVIEW

**Review by timja — APPROVED** [review]
works nicely

**Review by NotMyFault — APPROVED** [review]

**Review by Kevin-CB — APPROVED** [review]
Looks fine from a security perspective!

File: core/src/main/java/hudson/Functions.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Functions.java:2460]: As the Jellydoc is misleading, this would benefit from a Javadoc explaining use.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Is there a way to demonstrate the CMD/CTRL behavior? Right now it's unused code.

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/keyboard-shortcut.jelly:39]: It does not seem to be possible to define empty localizations. An attempt to do so will just use the default locale:

Screenshot 2023-08-29 at 14 29 18

**Review by daniel-beck — CHANGES_REQUESTED** [review]

**Comment by daniel-beck** [comment]
Subjective design feedback: The animation doesn't look good at the actual size. I first thought it was a bug, in particular on the default (non-dark) theme.

https://github.com/jenkinsci/jenkins/assets/1831569/9eadace4-adf3-4040-95f4-56ad5dac0dd8

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/keyboard-shortcut.jelly:39]: Good spot - updated.

File: core/src/main/java/hudson/Functions.java
Comment (janfaracik) [reply core/src/main/java/hudson/Functions.java:2479]: Added.

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/keyboard-shortcut.jelly:35]: Updated to reflect this.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Still doesn't work for me.
Screenshot 2023-12-06 at 14 49 09
(The build claims to be 2.434-SNAPSHOT so I'm not running the old one.)

Did you test this before resubmitting for review? If you did, what does the localized file look like?

**Comment by timja** [comment]
> Still doesn't work for me. Screenshot 2023-12-06 at 14 49 09 (The build claims to be 2.434-SNAPSHOT so I'm not running the old one.)
> 
> Did you test this before resubmitting for review? If you did, what does the localized file look like?

Works fine for me, there's a prefix and a suffix you need to update, I just fixed up a localisation (e.g. `pl`) as an example.

image

The animation itself isn't working for me right now, it might be a local issue but it's just a box. Would be nice if someone else could check it.

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2484]:
```suggestion
     * @return the translated shortcut, e.g. CMD+K to ⌘+K for macOS, CTRL+K for Windows
```

**Comment by daniel-beck** [comment]
> Works fine for me, there's a prefix and a suffix you need to update, I just fixed up a localisation (e.g. `pl`) as an example.

The point is that it needs to be possible to have one of the values be empty (since the approach of two sentence fragments, rather than one sentence with placeholder is used here). The part after "Press" is a complete sentence is German, starting with the key combination. `Press=` doesn't work.

**Comment by timja** [comment]
What about https://github.com/jenkinsci/jenkins/pull/8435/commits/5551fc91f2ec11e3115249223f79b3b0d2bd8526 @daniel-beck?

**Comment by daniel-beck** [comment]
Seems like a weird hack we should try not to do.

Best I could come up with myself is https://github.com/jenkinsci/jenkins/commit/4bd312d599a3127a536f0550f542ca284ffc3bae and that is only marginally better, as it requires escaping the placeholder (and requires leaving out the `<span>`s, but at least those seem to exist for no real reason).

**Comment by daniel-beck** [comment]
Looks like
```
Press=' '
```
would do it, by forcing a single space character. That seems cleaner than either of the above, except it would still produce an almost empty `<span>` unless specifically handled.

**Comment by timja** [comment]
@daniel-beck I like your single string approach, you happy with this?

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/keyboard-shortcut.jelly:29]:
```suggestion
      Tooltip message for the shortcut. Use the placeholder <code>{0}</code> escaped (wrapped in <code>&quot;</code>). A minimal message consisting of only the shortcut would be <code>&quot;{0}&quot;</code>.
```

(Assuming `<code>` works here.)

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/keyboard-shortcut.jelly:1]: It's weird that this entire file exists for the _tooltip_, yet its name doesn't indicate that that is the case. I would expect all of https://github.com/jenkinsci/jenkins/pull/8435/files#diff-3c611ba954e41706302b426f7b7669a8984d1302144d1b23e9bd1fa3a4cce6c2R62-R68 to be part of this control and I don't understand why it is not.

**Review by daniel-beck — DISMISSED** [review]
I didn't exactly provide a complete solution 😅 

English localization is missing, so hovering just shows `shortcut`.

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (timja) [reply core/src/main/resources/lib/layout/keyboard-shortcut.jelly]: it renders the code and &quot literally with the intellij plugin, unsure about stapler-maven-plugin itself, searching the code I couldn't see any handling of it though.

File: core/src/main/resources/lib/layout/keyboard-shortcut.jelly
Comment (timja) [reply core/src/main/resources/lib/layout/keyboard-shortcut.jelly:1]: Possibly because of `jenkins-search__shortcut` there's styling attached to it.

I've got a patch locally that mostly cleans it up but the CSS would need adapting.

Thoughts @janfaracik ?

```diff
diff --git a/core/src/main/resources/lib/layout/keyboard-shortcut.jelly b/core/src/main/resources/lib/layout/keyboard-shortcut.jelly
index 209ffabcf9..30711ecd93 100644
--- a/core/src/main/resources/lib/layout/keyboard-shortcut.jelly
+++ b/core/src/main/resources/lib/layout/keyboard-shortcut.jelly
@@ -23,21 +23,31 @@ THE SOFTWARE.
 -->
 
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define">
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout">
   <st:documentation>
     <st:attribute name="message">
-      Tooltip message for the shortcut. Use the placeholder {0} escaped (wrapped in '). A minimal message consisting of only the shortcut would be '{0}'.
+      Tooltip message for the shortcut. Use the placeholder {0} escaped (wrapped in '). A minimal message consisting of
+      only the shortcut would be '{0}'.
+    </st:attribute>
+    <st:attribute name="icon">
+      TODO
     </st:attribute>
     Displays a keyboard shortcut chip based on the component's body.
     CMD will automatically be translated to the user's operating system's appropriate variant.
   </st:documentation>
 
-  <div class="jenkins-tooltip__keyboard-shortcut">
-    <j:set var="body">
-      <div class="jenkins-tooltip__keyboard-shortcut__symbol">
-        <d:invokeBody trim="true" />
-      </div>
-    </j:set>
-    <j:out value="${h.formatMessage(message, h.translateModifierKeysForUsersPlatform(body))}" />
+  <j:set var="tooltip">
+    <div class="jenkins-tooltip__keyboard-shortcut">
+      <j:set var="body">
+        <div class="jenkins-tooltip__keyboard-shortcut__symbol">
+          <l:icon src="${attrs.icon}"/>
+        </div>
+      </j:set>
+      <j:out value="${h.formatMessage(message, h.translateModifierKeysForUsersPlatform(body))}"/>
+    </div>
+  </j:set>
+
+  <div class="jenkins-tooltip__keyboard-shortcut__container" data-html-tooltip="${tooltip}">
+    <l:icon src="${attrs.icon}"/>
   </div>
 </j:jelly>
diff --git a/core/src/main/resources/lib/layout/search-bar.jelly b/core/src/main/resources/lib/layout/search-bar.jelly
index 1e77c2be8a..ad94bf0598 100644
--- a/core/src/main/resources/lib/layout/search-bar.jelly
+++ b/core/src/main/resources/lib/layout/search-bar.jelly
@@ -58,14 +58,7 @@ THE SOFTWARE.
              spellcheck="false"
              disabled="${attrs.enabled == false ? true : null}" />
       <j:if test="${attrs.hasKeyboardShortcut != 'false'}">
-        <j:set var="tooltip">
-          <l:keyboard-shortcut message="${%shortcut}">
-            <l:icon src="symbol-search-shortcut" />
-          </l:keyboard-shortcut>
-        </j:set>
-        <div class="jenkins-search__shortcut" data-html-tooltip="${tooltip}">
-          <l:icon src="symbol-search-shortcut" />
-        </div>
+        <l:keyboard-shortcut message="${%shortcut}" icon="symbol-search-shortcut" />
       </j:if>
     </div>
   </div>

```

Comment (timja) [reply core/src/main/resources/lib/layout/keyboard-shortcut.jelly:1]: ping @janfaracik reminder about this

Comment (janfaracik) [reply core/src/main/resources/lib/layout/keyboard-shortcut.jelly:1]: Big thanks for this - I've updated it with this with some slight tweaks, mainly it also now supports text as well as symbols.

**Comment by janfaracik** [comment]
@daniel-beck @timja Updated this - big thanks to both of you for the comments and assistance.

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2587]:
```suggestion
        StaplerRequest2 currentRequest = Stapler.getCurrentRequest();
```

**Review by timja — APPROVED** [review]
LGTM - tested the latest changes and reviewed the code, much more in line with what was expected and cleaner, thanks. Apologies for the delay

**Comment by timja** [comment]
Re-requested @NotMyFault review as the code has changed a fair bit since last approved.

@daniel-beck do you have an ETA / plan to review again?

Thanks

**Comment by timja** [comment]
cc @NotMyFault ^^

**Review by krisstern — APPROVED** [review]
Makes sense to me

**Comment by krisstern** [comment]
@daniel-beck Would you like to review again? 

c.c. @NotMyFault

**Comment by timja** [comment]
> @daniel-beck Would you like to review again?

been 4 months since the original request for another review time to move on unless @daniel-beck you're able to make a commitment to review?

In the meantime:

-----

/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
